devito/ir/iet/scheduler.py

```diff
@@ -204,6 +204,9 @@ def push_object_on_stack(self, scope, obj):
         """Define an Array or a composite type (e.g., a struct) on the stack."""
         handle = self.stack.setdefault(scope, OrderedDict())
 
+        if obj in flatten(self.stack.values()):
+            return
+
```

Comment:
```
if 
    <LocalObject X>
    ...
else
    <LocalObject X>
```

> with this patch, X would only be declared once
We have a big challenge!

Reply:
yeah now u get it ? perhaps we should add an MFE to cover this case too, and then think about a proper rearchitecture

Reply:
For now, I think creating an `def push_array_on_stack` will avoid this problem.  

> and then think about a proper rearchitecture

Ok.

Reply:
```
if 
    <Array X>
    ...
else
    <Array X>
```
Can we assume that above case doesn't occur for array?

Reply:
mmhh... I'm fine with the current solution, but I'd also comment the code such that it points out that the above is unsupported 

in an ideal world, with this PR you would also:

- Add an MFE to test_ir.py::TestIETConstruction, marked as [xfail](http://doc.pytest.org/en/latest/skipping.html), that triggers the issue
- File an issue on github that explains the situation, the current implementation, and points to the relevant MFE

Reply:
Ok. I'm working on that.

Reply:
Hey @FabioLuporini 

The declarations are working fine, as seen at: `test_conditional_declarations`

Reply:
good, but what I meant is that we would need a test for `core` devito, since you're changing `core` devito here

Reply:
Ok, my bad.

Reply:
`test_conditional_declarations` created into `test_operator.py`
But I keep into `test_ops.py` too, because `test_operator.py` has `skipif(['yask', 'ops'])`, and the Array used in the backend ops comes from `devito/ops/types`.